File: Solutions/PalindromeNumber.cpp
// Leetcode Problem #9: https://leetcode.com/problems/palindrome-number
// Source: https://leetcode.com/problems/palindrome-number/solution

// Solution 1: Revert Half the Number | O(log(n)) time, O(1) space
class Solution {
public:
    bool isPalindrome(int x) {
        // x is not a palindrome if it is negative or if the last digit is 0 but x != 0
        if (x < 0 || (x % 10 == 0 && x != 0)) return false;
        
        // Loop through half the digits in x
        int reversed = 0;
        while (reversed < x) {
            // Append the last digit from x to reversed and remove it from x
            reversed = reversed * 10 + x % 10;
            x /= 10;
        }
        
        // If length of x is odd, reversed / 10 will remove the middle digit and check if its a palindrome
        return (x == reversed || x == reversed / 10);
    }
};

File: Solutions/PalindromeNumber.py
# LC #9: https://leetcode.com/problems/palindrome-number/

# Solution 1 (Revert half the number) | O(n) time, O(1) space
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # A negative number can't be a palindrome
        if x < 0: 
            return False
        
        # Reverse the last half of the number 
        reversedHalf = 0
        while x > reversedHalf:
            # Append the last digit from x and remove it from x
            reversedHalf = reversedHalf * 10 + (x % 10)
            x /= 10
        
        # If length of x is odd, reversedHalf will have an extra digit, so we do reversed / 10 to remove it
        # Since the middle digit doesn't matter in a palindrome, it's ok to just remove it
        return x == reversedHalf or x == reversedHalf / 10
        
